### PR TITLE
This turns the current state save feature to a menu

### PR DESCRIPTION
**Pull request by npwoods**
This turns the current state save feature to a menu.  One can still press 0-9, but you can also browse a menu when loading and saving state

### CODE REVIEW

**Comment by Tafoid** [comment]
You mentioned 0-9 can be pressed, but what of the other keyboard keys.  As it stand right now, you can use a-z and some others to load/save via F7/SHIFT+F7

File: src/frontend/mame/ui/state.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+
+//-------------------------------------------------
+//  itemref_from_slot_number
+//-------------------------------------------------
+
+void *menu_load_save_state_base::itemref_from_slot_number(int slot)
+{
+	return (void *)(size_t)slot;
```
Comment (cuavas) [inline src/frontend/mame/ui/state.cpp]: Casting from a signed integer to an unsigned integer of a different size has undefined behaviour.  Also, `size_t` is not guaranteed to be pointer-sized (that's what `std::uintptr_t` is for), and C-style casts are less readable than function-style casts.

**Comment by cuavas** [comment]
This is completely flawed - it's conflating keypresses with character input.  Hence it will never work with languages that use IME edit or Android.  Also as Tafoid says we support save/load on many more keys than 0-9, including joystick buttons.  You'll need to not lose that functionality.

File: src/frontend/mame/ui/state.cpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+		// stat the resulting file
+		auto entry = stat_searchpath(file_name, machine().options().state_directory());
+
+		// get the time as a local time string
+		char time_string[128];
+		if (entry != nullptr)
+		{
+			auto last_modified = std::chrono::system_clock::to_time_t(entry->last_modified);
+			std::strftime(time_string, sizeof(time_string), "%#c", std::localtime(&last_modified));
```
Comment (cuavas) [inline src/frontend/mame/ui/state.cpp]: %c is not loacale-aware or localisable.  Please use something localisable.

**Comment by rb6502** [comment]
Nathan, do you have any plans to fix this as @cuavas suggested, or should it be closed?  Personally I like the feature, so I was kind of looking forward to it :)

**Comment by cuavas** [comment]
We're slowly working towards it, but it's damn slow work untangling the UI.  Can we leave it open since it's on the roadmap?

**Comment by npwoods** [comment]
While the principle may be good, I'm not convinced that this PR needs to stay open.  I suspect that when the menuing system gets better, the work done here will in practice need to be done pretty much from scratch.  This is especially true given that my incorrect premise that we had a finite (i.e. - ten) list of slots, numbered from 0-9.

**Comment by npwoods** [comment]
Reopening this PR.  I've revamped it so that instead of a fixed list of slots (from 0-9), it now enumerates the state directory, listing files in reverse order of date.

**Pull request merged**